### DISCUSSION TITLE
How does fastapi record response time

### DISCUSSION THREAD

**Discussion by L1uNan** [fastapi/fastapi#8889] [Questions] [answered]
How does fastapi record interface response time and reflect it in its own API documents?

Is it possible to display the response time of the interface request on the openapi? 

Thank you very much.

**Comment by aminPial** [comment]
Not sure what you actually meant. Do you want to measure the time in between `processing a request and then response`? If so then do the following:



```python3
# check more on https://fastapi.tiangolo.com/tutorial/middleware/
import time
from fastapi import FastAPI, Request

app = FastAPI()

@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    start_time = time.time()
    response = await call_next(request)
    print("Time took to process the request and return response is {} sec".format(time.time() - start_time))
    return response
```
If this is not what you want, please explain a bit more about your query. Cheers.

**Comment by eignerfr** [comment]
thanks for the answer. would be cool to have the process time automatically in the log included (for instance in the end: 0.8sec)
2021-09-16 14:53:12,774 - INFO - 127.0.0.1:52962 - "GET / HTTP/1.1" 200 0.8
I'm pretty new to fastapi. I thought it should not be too difficult. Does someone know how to achieve this?

**Reply by Nabil372** [reply]
Check this out:

https://medium.com/@roy-pstr/fastapi-server-errors-and-logs-take-back-control-696405437983

**Reply by yaroslav-gwit** [reply]
These 2 articles combined made wonders for me:
https://medium.com/@roy-pstr/fastapi-server-errors-and-logs-take-back-control-696405437983
https://medium.com/1mgofficial/how-to-override-uvicorn-logger-in-fastapi-using-loguru-124133cdcd4e

In the end, you get a great logging solution, that looks like so (you do need to make some modifications, but they are trivial):

```log
INFO     2023-10-25 19:20:41.379 id: FastAPI.hoster - uvicorn.server:serve - Started server process [3018352]
INFO     2023-10-25 19:20:41.380 id: FastAPI.hoster - uvicorn.server:serve - Started server process [3018352]
INFO     2023-10-25 19:20:41.380 id: FastAPI.hoster - uvicorn.lifespan.on:startup - Waiting for application startup.
INFO     2023-10-25 19:20:41.380 id: FastAPI.hoster - uvicorn.lifespan.on:startup - Waiting for application startup.
INFO     2023-10-25 19:20:41.380 id: FastAPI.hoster - uvicorn.lifespan.on:startup - Application startup complete.
INFO     2023-10-25 19:20:41.380 id: FastAPI.hoster - uvicorn.lifespan.on:startup - Application startup complete.
INFO     2023-10-25 19:20:55.918 id: FastAPI.hoster - log_middleware:log_request_middleware - 127.0.0.1:0 - "GET /api/v1/vm/info-all-async" - 200 OK - 264.86ms
INFO     2023-10-25 19:20:57.212 id: FastAPI.hoster - log_middleware:log_request_middleware - 127.0.0.1:0 - "GET /api/v1/vm/info-all-async" - 200 OK - 451.80ms
INFO     2023-10-25 19:20:58.247 id: FastAPI.hoster - log_middleware:log_request_middleware - 127.0.0.1:0 - "GET /api/v1/vm/info-all-async" - 200 OK - 317.54ms
INFO     2023-10-25 19:21:15.325 id: FastAPI.hoster - log_middleware:log_request_middleware - 127.0.0.1:0 - "GET /api/v1/vm/info-all-async" - 200 OK - 250.24ms
INFO     2023-10-25 19:21:16.461 id: FastAPI.hoster - log_middleware:log_request_middleware - 127.0.0.1:0 - "GET /api/v1/vm/info-all-async" - 200 OK - 212.94ms
INFO     2023-10-25 19:21:19.921 id: FastAPI.hoster - log_middleware:log_request_middleware - 127.0.0.1:55764 - "GET /api/v1/health/default" - 200 OK - 0.36ms
INFO     2023-10-25 19:21:20.994 id: FastAPI.hoster - log_middleware:log_request_middleware - 127.0.0.1:55764 - "GET /api/v1/health/default" - 200 OK - 0.34ms
INFO     2023-10-25 19:21:21.233 id: FastAPI.hoster - log_middleware:log_request_middleware - 127.0.0.1:55764 - "GET /api/v1/health/default" - 200 OK - 0.26ms
INFO     2023-10-25 19:21:24.524 id: FastAPI.hoster - log_middleware:log_request_middleware - 127.0.0.1:55764 - "GET /api/v1/health/default" - 200 OK - 0.33ms
```

And thanks to `loguru` your logs are automatically redirected to a file, and rotated as needed.

**Comment by so2bin** [comment]
Has any resolution?

**Comment by LiberiFatali** [comment]
> thanks for the answer. would be cool to have the process time automatically in the log included (for instance in the end: 0.8sec) 2021-09-16 14:53:12,774 - INFO - 127.0.0.1:52962 - "GET / HTTP/1.1" 200 0.8 I'm pretty new to fastapi. I thought it should not be too difficult. Does someone know how to achieve this?

Could someone share some ideas about this? Thanks

**Comment by JarroVGIT** [comment, marked as answer]
You would need to override the logging produced by Uvicorn. Maybe this [blog](https://medium.com/1mgofficial/how-to-override-uvicorn-logger-in-fastapi-using-loguru-124133cdcd4e) will give you some idea on how to achieve this.

**Comment by kangig94** [comment]
```python
from uvicorn.config import LOGGING_CONFIG

app = FastAPI(...)

if __name__ == '__main__':
    LOGGING_CONFIG["formatters"]["access"]["fmt"] = (
        "%(asctime)s " + LOGGING_CONFIG["formatters"]["access"]["fmt"]
    )
    uvicorn.run('main:app', port=8000, reload=True)

# log
# 2023-02-24 11:01:06,936 INFO:     127.0.0.1:42006 - "GET / HTTP/1.1" 200 OK
```

Modify `LOGGING_CONFIG` in this way.
